### DISCUSSION TITLE
Suspend to RAM & Suspend-to-Idle are disabled

### DISCUSSION THREAD

**Discussion by sjuk** [fwupd/fwupd#7984] [Q&A] [answered]
**Describe the question**
Why is Suspend-to-idle disabled after I've explicitly set it to be the default variant?

After setting the Suspend-to-Idle as the default memory sleep mode...
...the following test passed.
```shell
✔ Suspend-to-ram:                Disabled: https://fwupd.github.io/libfwupdplugin/hsi.html#org.fwupd.hsi.SuspendToRam
```
...but the following test failed.
```shell
✘ Suspend-to-idle:               Disabled: https://fwupd.github.io/libfwupdplugin/hsi.html#org.fwupd.hsi.SuspendToIdle
```

**How Suspend-to-Idle has been set as the default memory sleep mode**
```shell
sudoedit /etc/default/grub
  replace: GRUB_CMDLINE_LINUX_DEFAULT="quiet splash"
  by: GRUB_CMDLINE_LINUX_DEFAULT="quiet splash mem_sleep_default=s2idle"
  save and close
sudo update-grub
sudo reboot
```
**How to check the currently set default memory sleep mode**
[System Sleep States](https://www.kernel.org/doc/html/latest/admin-guide/pm/sleep-states.html)
```shell
cat /sys/power/state 
freeze mem
cat /sys/power/mem_sleep 
[s2idle] deep
```
The following sleep states are supported by the kernel.
- freeze (suspend-to-idle)
- mem (The “mem” string is interpreted in accordance with the contents of file mem_sleep)

The following suspend variants are supported by the system.
- [s2idle] (Suspend-To-Idle, set to be the default one)
- deep (Suspend-To-Ram)

**fwupd version information**

```shell
fwupdmgr --version
compile   com.hughsie.libxmlb           0.3.18
compile   com.hughsie.libjcat           0.2.0
compile   org.freedesktop.fwupd         1.9.24
runtime   org.freedesktop.fwupd-efi     1.4
compile   org.freedesktop.gusb          0.4.8
runtime   com.hughsie.libxmlb           0.3.x
runtime   com.hughsie.libjcat           0.2.0
runtime   org.freedesktop.gusb          0.4.8
runtime   org.kernel                    6.8.0-47-generic
runtime   org.freedesktop.fwupd         1.9.24
```

**Comment by superm1** [comment]
Your _platform_ needs to assert that default from the ACPI tables.

**Reply by sjuk** [reply]
I wonder if there is anything I can do to improve the situation.

**Reply by superm1** [reply]
Some BIOSes offer a setting to control the default.  I know Lenovo does this for example.  If there is no setting then there is nothing you can do.

**Reply by sjuk** [reply]
I've asked the manufacturer of the motherboard for help.

**Reply by sjuk** [reply]
I've removed "mem_sleep_default=s2idle" from "GRUB_CMDLINE_LINUX_DEFAULT".
After that the Suspend-to-ram went back to enabled.
```Shell
✘ Suspend-to-ram:                Enabled
```
I wonder if the kernel command line does have any effect to the ACPI tables.
How does fwupd test for Suspend-to-ram and Suspend-to-idle?

**Reply by hughsie** [reply]
Does https://github.com/fwupd/fwupd/blob/main/plugins/linux-sleep/fu-linux-sleep-plugin.c help?

**Reply by sjuk** [reply]
@hughsie Thanks. This may clarify the first piece of the puzzle. Looks like it checks for  "[deep]" in "/sys/power/mem_sleep".
"deep" = Suspend-to-ram and the square brackets means it's the currently set default mode.
This is the reason why the Suspend-to-ram check resulted with disabled after switching the default to "s2idle".
**But why does the check for "s2idle" resulted with disabled when "/sys/power/mem_sleep" containts "[s2idle]"?**
Following the logic from fu-linux-sleep-plugin.c the Suspend-to-idle result is expected to be enabled.
This logic does not appear to be consistent for Suspend-to-ram (deep) and Suspend-to-idle (s2idle).

@superm1 FYI: fwupd appears not to check the ACPI tables, but "/sys/power/mem_sleep".
The statement "Your platform needs to assert that default from the ACPI tables." may generally be right.
Please forgive me the directness, but in the context of the question it maybe misleading.

**Reply by superm1** [reply]
I think that @hughsie only shared half of the tests.  Here's the other half.
https://github.com/fwupd/fwupd/blob/main/plugins/acpi-facp/fu-acpi-facp-plugin.c#L19

**Reply by sjuk** [reply]
Thanks for adding more information. To be honest I don't understand the other halft. Maybe there is more code involved and I'm not used to read and understand c code. Maybe you can spend some more words to explain why the test fails although "[s2idle]" is set as default in "/sys/power/mem_sleep".

**Reply by superm1** [reply]
The other half is pulling the FADT (called FACP for `$reasons`) and parsing it.  You can also do the same thing yourself using acpica-tools.  It knows about this offset too.
Here's how you can check it.  I just ran this on a system on my desk.

```sh
$ sudo cp /sys/firmware/acpi/tables/FACP .
$ sudo iasl -d FACP
$ grep Idle -i FACP.dsl
                      Low Power S0 Idle (V5) : 1
```

**Comment by sjuk** [comment]
Thanks. I'm able to reproduce it on my system. The result is "Low Power S0 Idle (V5) : 0".
Irritating is still the fact that Suspend-to-ram is tested by checking the currently default in "/sys/power/mem_sleep".
And Suspend-to-idle is tested by checking ACPI tables.
Does "Low Power S0 Idle (V5) : 0" in the ACPI tables mean that the system does not support s2idle?
Or is it only disabled (by default)? If yes, why is it available in "/sys/power/mem_sleep" and can be set as default?
Shouldn't it be more consistent?
Checking ACPI tables for both suspend modes **or** checking "/sys/power/mem_sleep" for both suspend modes.

**Reply by superm1** [reply]
Really IMO the S3 check is checking whether the kernel is setting the policy accordingly and the S2idle check is checking whether the firmware supports it.

You might reference the ACPI spec for more on the bit that is used.
https://uefi.org/htmlspecs/ACPI_Spec_6_4_html/05_ACPI_Software_Programming_Model/ACPI_Software_Programming_Model.html#fixed-acpi-description-table-fixed-feature-flags

It calls this `LOW_POWER_S0_IDLE_CAPABLE`.

**Comment by sjuk** [comment]
"A one informs OSPM that the platform is able to achieve power savings in S0 similar to or better than those typically achieved in S3. In effect, when this bit is set it indicates that the system will achieve no power benefit by making a sleep transition to S3."
It's a pity. it only states what a value of "one" means, but not what a value of "zero" means.
For now I stop spending more energy into that topic and accept that there is nothing I personally can do to improve the situation.
Maybe the vendor of the motherboard will be able.

**Reply by superm1** [reply, marked as answer]
Yeah; there is a lot of implications for the value this bit reflects both in Windows and Linux.  We shouldn't be changing the meaning of it in fwupd, the focal area absolutely should be for getting firmware that sets it.

**Reply by sjuk** [reply]
FYI: I've asked the manufacturer of the motherboard for help (on 28th October 2024).

**Reply by sjuk** [reply]
Answer from the manufacturer of the motherboard (Gigabyte): "We regret to inform you that due to the age of the motherboard, our team is no longer able to develop or maintain new BIOS versions. We ask for your understanding."
